### ISSUE TITLE
mesh11sd spams log on deactivated wifi interfaces

### ISSUE THREAD

**Issue by rossiniscarface** [openNDS/mesh11sd#7]
Hi,
i installed mesh11sd on my router which has two wifi devices (5g and 2.4g).
The 2.4g has one mesh defined nothing else. The 5g is deactivated.

The result is that mesh11sd writes all the time in the logs following:

`Thu Jun 30 12:27:48 2022 daemon.notice mesh11sd[17048]: wlan1 is not up - giving up for now.
Thu Jun 30 12:28:10 2022 daemon.notice mesh11sd[17048]: wlan1 is not up - giving up for now.
Thu Jun 30 12:28:32 2022 daemon.notice mesh11sd[17048]: wlan1 is not up - giving up for now.
Thu Jun 30 12:28:53 2022 daemon.notice mesh11sd[17048]: wlan1 is not up - giving up for now.
Thu Jun 30 12:29:15 2022 daemon.notice mesh11sd[17048]: wlan1 is not up - giving up for now.
Thu Jun 30 12:29:36 2022 daemon.notice mesh11sd[17048]: wlan1 is not up - giving up for now. `

I would suggest a check of the interfaces if they are a mesh point. 
Or some config options which interfaces should get monitored.

**Comment by rossiniscarface** [comment]
Using openwrt snapshot with mesh11sd 1.1.1-2

`root@rectangle:~#uci show wireless
wireless.radio0=wifi-device
wireless.radio0.type='mac80211'
wireless.radio0.hwmode='11g'
wireless.radio0.path='1e140000.pcie/pci0000:00/0000:00:01.0/0000:02:00.0'
wireless.radio0.cell_density='0'
wireless.radio0.country='AT'
wireless.radio0.channel='11'
wireless.radio0.htmode='HT20'
wireless.radio1=wifi-device
wireless.radio1.type='mac80211'
wireless.radio1.hwmode='11a'
wireless.radio1.path='1e140000.pcie/pci0000:00/0000:00:00.0/0000:01:00.0'
wireless.radio1.cell_density='0'
wireless.radio1.txpower='20'
wireless.radio1.country='AT'
wireless.radio1.htmode='VHT80'
wireless.radio1.channel='36'
wireless.radio1.disabled='1'
wireless.wifinet2=wifi-iface
wireless.wifinet2.device='radio0'
wireless.wifinet2.mode='mesh'
wireless.wifinet2.encryption='sae'
wireless.wifinet2.mesh_id='xxx-lan-mesh'
wireless.wifinet2.mesh_fwding='1'
wireless.wifinet2.mesh_rssi_threshold='-80'
wireless.wifinet2.key='xxx'
wireless.wifinet2.network='lan'
wireless.wifinet2.mesh_gate_announcements='1'
wireless.wifinet2.ifname='mesh0'
`
`
root@rectangle:~#iw dev
phy#1
        Interface wlan1
                ifindex 10
                wdev 0x100000001
                addr b0:39:56:22:e6:5f
                type managed
                txpower 5.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt  hashcol tx-bytes        tx-packets
                        0       0       0       0       0       0phy#0
        Interface mesh0
                ifindex 29
                wdev 0x8
                addr b0:39:56:22:e6:5e
                type mesh point
                channel 11 (2462 MHz), width: 20 MHz, center1: 2462 MHz
                txpower 20.00 dBm
                multicast TXQ:
                        qsz-byt qsz-pkt flows   drops   marks   overlmt  hashcol tx-bytes        tx-packets
                        0       0       55481   0       0       18359528         57529
`
No mesh11sd status as i deinstalled it. But doesn‘t matter as it‘s clear in the code that it never checks if the type of wifi interface is a mesh.

**Comment by bluewavenet** [comment]
@rossiniscarface 

> it‘s clear in the code that it never checks if the type of wifi interface is a mesh.

What code did you look in? /s

For each interface in turn, mesh11sd tests:
1. Is the interface up?
2. If it is not up, log to the syslog and skip to the next interface.
3. If it is up, is it a mesh interface?

Your output of `iw dev` shows phy#1 has two enabled interfaces, wlan1 and mesh0.
But you do not have a wifi-iface section for wlan1 in the config.
This means it is configured with basic default values to be an enabled virtual interface of phy#1, but not enough configuration  for it to come up.

The problem you have is that you have deleted the wifi-iface config section you wanted to disable and otherwise edited the rest of the config file.
Removing the wifi-iface sections only removes options from the config, **it does not disable the interface**.
You have added the option to disable the 5GHz **radio** which will override any virtual wifi-iface enabling option for that radio.

It would be better to delete the config file and run `wifi config` to recreate the default config file.
Then disable the 5G radio as you did before.
1. Add `option disabled '1'` to the wifi-iface section for the 2GHz AP
2. Add the basic wifi-iface section for the 2GHz mesh virtual interface
3. Add `option disabled '0'` to the wifi-iface section for the 2GHz MESH

Restart the wifi using `wifi` command
Install mesh11sd and it will correctly name the mesh interface and add the required mesh parameters to the mesh network.
This either has to be done:
 * manually after the boot process has completed and again after every restart of the wireless subsystem
 or
 * by installing mesh11sd to monitor and maintain parameter settings

**Note** most mesh parameters cannot be configured in the wireless config, including mesh_gate_announcements.

**Comment by rossiniscarface** [comment]
I mean this line of code: 
[https://github.com/openNDS/mesh11sd/blob/4c4d9091bc559e49882903fcbd00aefa53b04f42/src/mesh11sd#L397](https://github.com/openNDS/mesh11sd/blob/4c4d9091bc559e49882903fcbd00aefa53b04f42/src/mesh11sd#L397)

Sorry, the output was a little malformed. There was a phy#0 in between. I corrected the formatting.

And i am sorry but my config is not the problem. Why even bother with an iw interface which isn‘t even configured as an mesh point? Why should mesh11sd inform me that the interface is down which shouldn‘t be up in the first place and isn‘t even configuread as a mesh?
It is totally fine for OpenWRT to delete the Interface and not the Physical device. 

Output of iwinfo: 

```
root@rectangle:~#iwinfo 
mesh0     ESSID: "xxx-lan-mesh"
          Access Point: xx:xx:xx:xx:xx:xx
          Mode: Mesh Point  Channel: 11 (2.462 GHz)
          Center Channel 1: 11 2: unknown
          Tx-Power: 20 dBm  Link Quality: 44/70
          Signal: -66 dBm  Noise: unknown
          Bit Rate: 130.0 MBit/s
          Encryption: WPA3 SAE (CCMP)
          Type: nl80211  HW Mode(s): 802.11bgn
          Hardware: 14C3:7603 14C3:7603 [MediaTek MT7603E]
          TX power offset: none
          Frequency offset: none
          Supports VAPs: yes  PHY name: phy0

wlan1     ESSID: unknown
          Access Point: xx:xx:xx:xx:xx:xx
          Mode: Client  Channel: unknown (unknown)
          Center Channel 1: unknown 2: unknown
          Tx-Power: 5 dBm  Link Quality: unknown/70
          Signal: unknown  Noise: unknown
          Bit Rate: unknown
          Encryption: unknown
          Type: nl80211  HW Mode(s): 802.11nac
          Hardware: 14C3:7615 7615:14C3 [MediaTek MT7615E]
          TX power offset: none
          Frequency offset: none
          Supports VAPs: yes  PHY name: phy1

```

**Comment by bluewavenet** [comment]
@rossiniscarface 

> wlan1     ESSID: unknown

So yes, iwinfo agrees with iw, a virtual interface "wlan1" has been created. 

> It is totally fine for OpenWRT to delete the Interface and not the Physical device.

No it is not totally fine, as indicated by both `iw` and `iwinfo` (and `ifconfig` and `ip link`) there is a broken virtual interface.....

> [mesh11sd/src/mesh11sd](https://github.com/openNDS/mesh11sd/blob/4c4d9091bc559e49882903fcbd00aefa53b04f42/src/mesh11sd#L397)
> 
> Line 397 in [4c4d909](https://github.com/openNDS/mesh11sd/commit/4c4d9091bc559e49882903fcbd00aefa53b04f42)

If you had read a few lines further on you would have found this:

```
					# get list of mesh parameters for this interface
					get_params

					if [ -z "$params" ]; then
						# this is not a mesh interface
						continue
					else
						# Check if this interface has a uci ifname .........
```

> And i am sorry but my config is not the problem.

Sorry, but it is, as indicated by the iw and iwinfo outputs you have provided.

Your incorrect config has resulted in an orphaned virtual interface which is an error.

If you wish to suppress mesh11sd's reporting of your error you can set debuglevel to "0". The mesh will _probably_ still work with this error condition......

**Comment by bluewavenet** [comment]
@rossiniscarface 

> why is mesh11sd messing with interfaces which doesn‘t belong to a mesh?

It does not "mess" with non-mesh interfaces. I checks all interfaces to see which ones are mesh.

> As you can see iwinfo, iw states clearly mode!=mesh 

Yes, in fact mesh11sd uses the iw method of checking for mesh.

>  Suppress reporting would also suppress errors for the real mesh interfaces. So this shouldn‘t also be the solution.

I agree. It also would be possible to suppress the error logging for this without suppressing other log events, but that is also not a solution.

The solution to this issue is to correct the invalid config that creates the error in the first place.

**Comment by rossiniscarface** [comment]
Or first check if an interface belongs to a mesh and then log if its up or down?

**Comment by bluewavenet** [comment]
@rossiniscarface 
There are numerous situations where a wireless virtual interface may take several minutes to come up eg 5GHz DFS when the driver thinks it may be detecting radar.
Until the interface is up, it is not possible to reliably determine in a generic way whether the interface is going to be configured for mesh when it is finally up. 

It is only optional to name the virtual interfaces in the uci config and without such naming a default name is generated, so peaking into the uci config file is not good enough either.

In addition, some people, or some packages, might decide to bypass the uci config and configure manually using `iw` directly.

> Or first check if an interface belongs to a mesh and then log if its up or down?

TL;DR If an interface is down, you cannot check if it is going belong to a mesh when it comes up.

**Comment by rossiniscarface** [comment]
Sorry i am still not convinced. If the interface clearly states mode!=mesh, the uci config etc … i don‘t get it why mesh11sd, which is clearly for mesh interfaces, checks if it is up and spams the log. 
Also your argument about manually created interfaces just refuted your argument about „fixing the config first“.
But fine for me. 
If you don‘t see a solution for this than you should close this issue.

**Comment by bluewavenet** [comment]
@rossiniscarface 
From your `iwinfo` output:
```
wlan1     ESSID: unknown
          Access Point: xx:xx:xx:xx:xx:xx
          Mode: Client  Channel: unknown (unknown)
```
"Mode: Client" here is the key.
After some testing to duplicate your configuration, I have come to the conclusion you have not divulged some part of your configuration, as iwinfo is reporting wlan1 as in client mode and this is not indicated anywhere in what you have shown.
_My hypotheisis on the origin of your virtual interface wlan1 is not correct._

Nevertheless, I concede your point that mesh11sd should only check if mesh interfaces are up.

I have attached mesh11sd v1.1.2beta that should differentiate between mesh and non-mesh interfaces.
I would appreciate it if you could test on your system.

[mesh11sd.txt](https://github.com/openNDS/mesh11sd/files/9027420/mesh11sd.txt)

Install mesh11sd v1.1.1 using opkg in the normal way.
Download v1.1.2beta from the link above.
Rename to "mesh11sd" from "mesh11sd.txt".
Copy to `/usr/sbin`, overwriting the existing file (assuming the package is already installed). Make sure it is flagged as executable once copied. Then run `service mesh11sd restart`.

**Comment by rossiniscarface** [comment]
Was able to successful testing your changes. 
No logging for non-mesh interfaces but still logging for mesh interfaces.
Nice done. 👍

**Comment by bluewavenet** [comment]
@rossiniscarface 
This is now merged into master for upcoming v1.2.0 release.
